### ISSUE TITLE
Translatable strings in HelpWML can't include a caret to be displayed to the user

### ISSUE THREAD

**Issue by stevecotton** [wesnoth/wesnoth#4359]
I want to show a literal caret character to the user:
```
#ifdef EDITOR
    # po: this hint is only shown in the map editor, the ^Uf, ^Tf and Tb are all terrain codes and should be left as-is
    help_topic_text= _ "editor^The overlays '^Uf' and '^Ufi' are deprecated because, although the graphics show an overlay with another terrain underneath, the movement and defense stats are purely fungus terrain, ignoring the underlying terrain. The recommended replacements are '^Tf' and '^Tfi' which look the same but have the stats implied by the graphics. For hexes where the stats of purely fungus are wanted, the new 'Tb' terrain is a mushroom base terrain, which can be used with or without '^Tf' on top."
#endif
```

The code in src/gettext.cpp for handling translatable strings uses the variable name `firsthat` when searching for the caret character, so I assume it's only meant to remove the initial `editor^` part of the string. However, when that terrain is viewed in the help browser the text is cut to only `Tf' on top.`

The text `editor^<italic>text='^Uf'</italic>` will simply cause the help window to close - it's being cut to `Uf'</italic>` and causing an error in the help browser's text parsing.

Edit: the help browser uses its own markup format, and this issue doesn't involve Pango.

**Comment by jostephd** [comment]
> The code in src/gettext.cpp for handling translatable strings uses the variable name `firsthat` when searching for the caret character, so I assume it's only meant to remove the initial `editor^` part of the string. However, when that terrain is viewed in the help browser the text is cut to only `Tf' on top.`

Change strrchr to strchr? No string in mainline has two `^`s, and I see nothing in https://wiki.wesnoth.org/TranslationsWML#Hidden_Text and https://wiki.wesnoth.org/GettextForWesnothDevelopers about intentionally making it impossible to use `^` in a translatable string.

> Although this text is rendered with Pango, entity codes such as `&#94;` are shown to the user as `&#94;` rather than converting them to the caret character.

`&lt;` for `<` does seem to work in Pango text, and for basically everything else we can just use the character literally, can't we? (Once we fix this bug, that is.)

> The text `editor^<italic>text='^Uf'</italic>` will simply cause the help window to close - I assume it's being cut to `Uf'</italic>` and causing a Pango error.

Odd, I can't replicate that in the tips of the day:

```
$ cat data/tips.cfg
#textdomain wesnoth

[tip]
    text= _ "Use the map editor’s <i>Areas</i> &lt;foo&gt; menu to define regions later addressed by event or AI coding."
    source= _ "editor^<italic>text='^Uf'</italic>"
[/tip]
```

![2019-09-18-084411_651x139_scrot](https://user-images.githubusercontent.com/24784687/65132452-8055dd00-d9f0-11e9-9f98-41d350b35861.png)

and I get `error gui/layout: pango_text::set_markup text 'Uf'</italic>' has broken markup, set to normal text.` on stderr.

**Comment by soliton-** [comment]
I agree that this just seems like a bug/oversight. Much better to disallow `^` in the translation hint than in the actual translatable string.

**Comment by stevecotton** [comment]
> Odd, I can't replicate that in the tips of the day:
> ...
> and I get `error gui/layout: pango_text::set_markup text 'Uf'</italic>' has broken markup, set to normal text.` on stderr.

Urgh, it seems the help browser has its own tag-parser:
https://github.com/wesnoth/wesnoth/blob/1fe32ff69d28f3bbefde68f6b4eae848576d35c0/src/help/help_impl.hpp#L272-L275
https://github.com/wesnoth/wesnoth/blob/1fe32ff69d28f3bbefde68f6b4eae848576d35c0/src/help/help_impl.cpp#L1154-L1158

**Comment by jostephd** [comment]
> Urgh, it seems the help browser has its own tag-parser:

The help doesn't use pango syntax.

**Comment by stevecotton** [comment]
Thanks for the correction, I've just checked and `&#94;` does work in areas that use Pango, so I'll update the title to say that this is only in HelpWML.

**Comment by stevecotton** [comment]
Yes - except it's that there's no way to escape `^` in HelpWML. Probably no way to escape `<` either...

**Comment by jostephd** [comment]
It's possible to escape `<` with a backslash:

https://github.com/wesnoth/wesnoth/blob/master/src/help/help_impl.cpp#L1129

So I think this takes care of both issues:

```
diff --git a/data/core/help.cfg b/data/core/help.cfg
index 6f7d5a81602..42b6beac314 100644
--- a/data/core/help.cfg
+++ b/data/core/help.cfg
@@ -97,10 +97,12 @@
     [topic]
         id=..introduction
         title= _ "Introduction"
-        text="<img>src=misc/logo-bg.png~BLIT(misc/logo.png) align=middle box=no</img>" + _ "
+        text="<img>src=misc/logo-bg.png~BLIT(misc/logo.png) align=middle box=no</img>" + _ "^
 
 <italic>text='Battle for Wesnoth'</italic> is a turn-based fantasy strategy game somewhat unusual among modern strategy games. While other games strive for complexity, <italic>text='Battle for Wesnoth'</italic> strives for simplicity of both rules and gameplay. This does not make the game simple, however — from these simple rules arise a wealth of strategy, making the game easy to learn but a challenge to master.
 
+foo \< bar ^ baz
+
 The following pages outline all you need to know to play Wesnoth. As you play, new information is added to the various categories as you come across new aspects of the game. For more detailed information on special situations and exceptions, follow the included links."
     [/topic]
     # wmllint: markcheck on
diff --git a/src/gettext.cpp b/src/gettext.cpp
index 5a5406c3d8c..af6feb4d6c3 100644
--- a/src/gettext.cpp
+++ b/src/gettext.cpp
@@ -405,7 +405,7 @@ std::string dsgettext (const char * domainname, const char *msgid)
 {
 	std::string msgval = dgettext (domainname, msgid);
 	if (msgval == msgid) {
-		const char* firsthat = std::strrchr (msgid, '^');
+		const char* firsthat = std::strchr (msgid, '^');
 		if (firsthat == nullptr)
 			msgval = msgid;
 		else
@@ -420,7 +420,7 @@ std::string dsngettext (const char * domainname, const char *singular, const cha
 	std::lock_guard<std::mutex> lock(get_mutex());
 	std::string msgval = bl::dngettext(domainname, singular, plural, n, get_manager().get_locale());
 	if (msgval == singular) {
-		const char* firsthat = std::strrchr (singular, '^');
+		const char* firsthat = std::strchr (singular, '^');
 		if (firsthat == nullptr)
 			msgval = singular;
 		else
```

Did I leave anything out?

**Comment by jostephd** [comment]
@stevecotton What are we waiting for? I assume we can just commit that patch and then you can add the editor help string from the OP. Are you waiting on me to commit that or something?

**Comment by stevecotton** [comment]
Sorry, yes I thought you'd commit it. I could commit it tomorrow (don't have time for testing tonight), along with the editor help string.

**Comment by jostephd** [comment]
Yes please! :) I didn't see your thumbs up until today (github doesn't generate a notification on that).

**Comment by stevecotton** [comment]
Fixed in #4412 and https://wiki.wesnoth.org/index.php?title=GettextForWesnothDevelopers&curid=1410&diff=64692&oldid=61925
